src/frontend/src/flows/addDevice/manage/verifyTentativeDevice.ts

```diff
@@ -1,14 +1,72 @@
 import { html, render } from "lit-html";
+import { unreachableLax, unknownToString } from "../../../utils/utils";
+import { createRef, ref } from "lit-html/directives/ref.js";
+import { asyncReplace } from "lit-html/directives/async-replace.js";
+import { withRef } from "../../../utils/lit-html";
 import { AuthenticatedConnection } from "../../../utils/iiConnection";
 import { withLoader } from "../../../components/loader";
-import { renderManage } from "../../manage";
 import { displayError } from "../../../components/displayError";
-import { DeviceData } from "../../../../generated/internet_identity_types";
-import { toggleErrorMessage } from "../../../utils/errorHelper";
-import { setupCountdown } from "../../../utils/countdown";
+import { Chan } from "../../../utils/utils";
+import { AsyncCountdown } from "../../../utils/countdown";
 import { mainWindow } from "../../../components/mainWindow";
+import { VerifyTentativeDeviceResponse } from "../../../../generated/internet_identity_types";
+
+// The type that we return, pretty much the result of the canister
+// except that all retries have been exhausted
+type VerifyResult =
+  | Exclude<VerifyTentativeDeviceResponse, { wrong_code: unknown }>
+  | { too_many_attempts: null };
+
+// A helper type for the page verification function where a retry may be possible/requested
+type VerifyResultOrRetry = VerifyResult | { retry: null };
+
+const verifyTentativeDeviceTemplate = ({
+  alias,
+  remaining,
+  verify,
+  doContinue,
+  cancel,
+}: {
+  alias: string;
+  remaining: AsyncIterable<string>;
+  cancel: () => void;
+  verify: (value: string) => Promise<VerifyResultOrRetry>;
+  doContinue: (result: VerifyResult) => void;
+}) => {
+  // The error shown on failed verification
+  const showRetryPrompt = new Chan<boolean>(false);
+  const retryPromptHidden = showRetryPrompt.map((showRetry) =>
+    showRetry ? "" : "is-hidden"
+  );
+
+  // The error shown on bad input (e.g. "")
+  const badInput = new Chan<boolean>(false);
+  const hasError = badInput.map((e) => (e ? "has-error" : ""));
+
+  const pinInput = createRef<HTMLInputElement>();
+
+  // When submitting, we either give visual feedback on bad input (e.g. "") or we
+  // use the provided verification function to decide whether to retry or contine by forwarding
+  // the result.
+  const submit = () => {
+    withRef(pinInput, async (pin) => {
+      const value = pin.value;
+      if (value === "") {
```

Comment:
nth: I would not mind if instead of "send(true|false)" it would be possible to make it more readable "send("ok"|"error"). got a bit lost while reading this snippet.